### PR TITLE
Write state using compact JSON representation

### LINKED ISSUE #1593: Pretty-printed JSON representation of a state increases file size by ~20-25%
### OpenTofu Version

```shell
1.8
```

### Use Cases

We have a few use cases when the state size is very big, and we were looking into optimizing performance. What I noticed is that by default, we're using pretty-printed JSON when storing the state, and the resulting file size is much bigger than if we use compact representation, and this make read/write of state slower, not taking into account increased storage and egress costs when using cloud backends.

For example, here is comparison of the state with ~70k objects - original state (`mkdirs.tfstate`) is converted into compact representation by running `jq -c .` on it.

```
84096 -rw-rw-r--@ 1 alex  staff  42996911 Apr 27 10:53 mkdirs-compact.tfstate
114112 -rw-r--r--@ 1 alex  staff  57680266 Apr 27 10:50 mkdirs.tfstate
```

### Attempted Solutions

.

### Proposal

Enable compact representation by default. I can make PR to support it if it's ok

### References

_No response_

### PR DESCRIPTION
**Pull request by alexott**
This PR changes the writing of the state using compact JSON  (non-indented) representation.  This significantly improves performance when working with huge states.

For example (on ~450k objects in the state):

* state size is decreased from 510.1MB to 386.6MB
* Time to write to S3 decreased from 112 seconds to 94 seconds
* Time to read from S3 decreased from 87 seconds -> 55 seconds

Resolves #1593 

## Target Release

1.8.0

## Checklist

- [ ] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [ ] I have not used an AI coding assistant to create this PR.
- [ ] I have marked any code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from. 

### Go checklist

<-- If your PR contains Go code, please make sure you check off all items on this list: --> 

- [ ] I have run golangci-lint on my change and receive no errors relevant to my code.
- [ ] I have run existing tests to ensure my code doesn't break anything.
- [ ] I have added tests for all relevant use cases of my code, and those tests are passing.
- [ ] I have only exported functions, variables and structs that should be used from other packages.
- [ ] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

File: internal/states/statefile/version4.go [hunk 1]
```diff
@@ -424,7 +425,13 @@ func writeStateV4(file *File, w io.Writer, enc encryption.StateEncryption) tfdia
 
 	sV4.normalize()
 
-	src, err := json.MarshalIndent(sV4, "", "  ")
+	var src []byte
+	var err error
+	if _, exists := os.LookupEnv("TF_USE_COMPACT_STATE_FORMAT"); exists {
```
[the remark this answers is not in the document]
Comment (alexott) [reply internal/states/statefile/version4.go:430]: I'm happy to make changes based on the guidance. Right now, I just check for presence of the environment variable to enable this behavior.

[the remark this answers is not in the document]
Comment (cam72cam) [reply internal/states/statefile/version4.go:430]: My gut is saying that this should be configured as part of the project, but we are just starting to look at supporting something like $PWD/.tofurc to allow per-project settings.  Thanks for your patience @alexott while we dig into this conversation we've been putting off for a while.

[the remark this answers is not in the document]
Comment (alexott) [reply internal/states/statefile/version4.go:430]: I'm really not sure what would be the best way to configure it - command-line flag, environment variable, or config file.

[the remark this answers is not in the document]
Comment (cube2222) [reply internal/states/statefile/version4.go:430]: In my opinion (and I think we should further discuss this till we reach consensus) this should not be configurable at all.

That is, in my opinion we should just switch to the compact representation and be done with it.

[the remark this answers is not in the document]
Comment (cam72cam) [reply internal/states/statefile/version4.go:430]: I agree with @cube2222, no one should be directly consuming the state file.  If someone ever needs to inspect it in depth, `cat terraform.tfstate | jq` or similar will suffice.

[the remark this answers is not in the document]
Comment (alexott) [reply internal/states/statefile/version4.go]: I've changed the PR to always use compact representation

**Review by cam72cam — DISMISSED** [review]

**Comment by alexott** [comment]
Tests are fixed

**Comment by cam72cam** [comment]
I had a small heart attack when looking at the LoC totals in the diff, before remembering that it was the whole goal of this PR :laughing:

**Review by cam72cam — APPROVED** [review]

File: internal/command/apply_test.go [hunk 2]
```diff
@@ -1594,7 +1594,7 @@ func TestApply_backup(t *testing.T) {
 				Name: "foo",
 			}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
 			&states.ResourceInstanceObjectSrc{
-				AttrsJSON: []byte("{\n            \"id\": \"bar\"\n          }"),
+				AttrsJSON: []byte("{\"id\":\"bar\"}"),
```
Comment (ollevche) [inline internal/command/apply_test.go:1597]:
```suggestion
				AttrsJSON: []byte(`{"id":"bar"}`),
```

**Review by ollevche — APPROVED** [review]

**Review by cam72cam — APPROVED** [review]

**Comment by eduzgun** [comment]
@cam72cam Some of the testdata in `/internal/states/statefile/testdata/roundtrip` looks wrong and is causing ` go test ./internal/states/statefile` tests to fail. These files are: v4-cbd.out.tfstate, v4-foreach.out.tfstate, v4-modules.out.tfstate, v4-simple.in.tfstate. Should I create a new issue for this?

**Comment by cam72cam** [comment]
@eduzgun, sure create an issue.  I can't reproduce that myself, the tests are passing on main.

**Comment by bobrik** [comment]
This is somewhat annoying for simple use-cases where a file is saved locally. Previously the changes could be reasonable diffed and saved into git, now it's a single gigantic line that cannot be viewed as a diff. It's evident just from the updated testdata itself:

* https://github.com/opentofu/opentofu/commit/522a859cf012ab2f1e2a42355281d2f7ad961c28#diff-e184610507824cd7d0b832b4e3b9267e88df76f63bd2b07e66c33157318ef6a0

Further changes will be even less readable.

I think there's a middle ground here, where a compact representation can be used for non-local backends (or perhaps if there is enough data for it to matter).

**Comment by cam72cam** [comment]
Could you create an issue linking to this PR @bobrik?  I'm curious what patterns you follow that makes a git diff of the state file useful.

**Comment by bobrik** [comment]
I made #1947.

**Pull request merged**